### PR TITLE
Type wrapper

### PR DESCRIPTION
**Pull request by pacrob**
### What was wrong?

Adding types to `wrapper.py`.

Rather than keeping 2 lists of accepted `geth_kwargs` (in the args to `construct_popen_command` and in the `GethKwargs` model), `construct_popen_command` now takes the normal `**geth_kwargs` as its arg, then uses `GethKwargs` to validate and fill the default values.

Branched from #204, merge that first

### Todo:

- [x] Clean up commit history
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/ethereum/py-geth/blob/main/newsfragments/README.md)

#### Cute Animal Picture

image

### CODE REVIEW

File: geth/utils/validation.py [hunk 1]
```diff
@@ -22,28 +22,20 @@ class GethKwargs(BaseModel):
...
         geth_executable: (str) The path to the geth executable
         ipcdisable (bool): Disable the IPC-RPC server.
         max_peers (str): Maximum number of network peers.
-        metrics (bool): Enable metrics collection.
         mine (bool): Enable mining.
         network_id (str): The network identifier.
         nodiscover (bool): Disable network discovery.
         password (str): Path to a file that contains a password.
         port (str): The port to listen on.
-        pprof (bool): Enable pprof collection.
```
Comment (kclowes) [inline geth/utils/validation.py:38]: I think this is still around?

File: geth/utils/validation.py [hunk 2]
```diff
@@ -22,28 +22,20 @@ class GethKwargs(BaseModel):
         autodag (bool): generate a DAG (pre-merge only)
         allowinsecure (bool): Allow insecure accounts.
         data_dir (str): The directory to store the chain data.
-        etherbase (str): The account to send mining rewards to.
-        extradata (str): Extra data to include in the block.
-        gasprice (int): Minimum gas price for mining.
-        genesis (str): The genesis file to use.
```
Comment (kclowes) [inline geth/utils/validation.py:28]: Is passing a genesis file no longer supported?

File: geth/utils/validation.py [hunk 3]
```diff
@@ -55,48 +47,40 @@ class GethKwargs(BaseModel):
...
     shh: bool | None = None
-    targetgaslimit: int | None = None
+    stdin: str | None = None
+    suffix_args: list[str] | None = None
+    suffix_kwargs: dict[str, Any] | None = None
     tx_pool_global_slots: int | None = None
     tx_pool_price_limit: int | None = None
     unlock: str | None = None
     verbosity: int | None = None
-    vmodule: str | None = None
```
Comment (kclowes) [inline geth/utils/validation.py:87]: I think this is still an option?

File: geth/utils/validation.py [hunk 4]
```diff
@@ -22,28 +22,20 @@ class GethKwargs(BaseModel):
         autodag (bool): generate a DAG (pre-merge only)
         allowinsecure (bool): Allow insecure accounts.
         data_dir (str): The directory to store the chain data.
-        etherbase (str): The account to send mining rewards to.
```
Comment (fselmo) [inline geth/utils/validation.py:25]: I'm going to list all the keys I did not see put back in here

- `etherbase`
- `targetgaslimit`
- `extradata`
- `genesis`
- `gasprice`
- `metrics`
- `pprof`
- `vmodule`

I'm kind of curious where these came from to begin with (and we missed, my fault included). They don't seem to be used anywhere, pre-type-hinting. I think everything else lines up with `construct_popen_command` though so I think this lgtm?

I see @kclowes had a similar comment so I'll wait until that's all resolved.

File: geth/utils/validation.py [hunk 5]
```diff
@@ -55,48 +47,40 @@ class GethKwargs(BaseModel):
...
-    extradata: str | None = None
-    gasprice: int | None = None
+    dev_mode: bool | None = False
     gcmode: Literal["full", "archive"] | None = None
-    genesis: str | None = None
+    geth_executable: str | None = None
     ipc_disable: bool | None = None
     ipc_path: str | None = None
     max_peers: str | None = None
-    metrics: bool | None = None
```
Comment (kclowes) [inline geth/utils/validation.py:67]: I think this is still around?

File: geth/wrapper.py [hunk 6]
```diff
@@ -116,185 +120,138 @@ def extend(self, value_list: Iterable[Any]) -> None:
...
             raise ValueError("`mine` must be truthy when specifying `miner_etherbase`")
-        builder.extend(("--miner.etherbase", miner_etherbase))
+        builder.extend(("--miner.etherbase", gk.miner_etherbase))
 
-    if autodag:
+    if gk.autodag:
         builder.append("--autodag")
 
-    if shh:
+    if gk.shh:
```
Comment (kclowes) [inline geth/wrapper.py:222]: This can go, I think

**Review by kclowes — COMMENTED** [review]
I couldn't 100% tell the rhyme or reason to delete some options but not others, so feel free to take or leave my comments 😆 typing looks good though. I think you're moving that `**geth_kwargs: Any` to something stricter?

**Comment by pacrob** [comment]
> I couldn't 100% tell the rhyme or reason to delete some options but not others, so feel free to take or leave my comments 😆 typing looks good though. I think you're moving that `**geth_kwargs: Any` to something stricter?

All these deletions are fields that copilot filled in when I was very first creating the GethKwargs model. In this PR, I went though all the kwargs that are actually used anywhere in `py-geth` and removed the ones that aren't, as there isn't a mechanism for applying additional fields directly (the user can add them via `suffix_args` and `suffix_kwargs` if they want to.)

We looked at using a `TypedDict` to validate the geth_kwargs dict, but to do it properly would break some existing function signatures. Once all the broader typing & cleanup is done, I'll bring proposed breaking changes and see how everyone feels.

File: geth/utils/validation.py (hunk 4 above)
Comment (pacrob) [reply geth/utils/validation.py:25]: co-pilot filled 'em in when I first created the GethKwargs model, just cleaning up/paring back to what we actually use.

**Review by fselmo — APPROVED** [review]
Co-pilot comment makes sense of the missing things to me. Lgtm 👍🏼

File: geth/utils/validation.py [hunk 7]
```diff
@@ -22,28 +22,20 @@ class GethKwargs(BaseModel):
         autodag (bool): generate a DAG (pre-merge only)
```
[the remark this answers is not in the document]
Comment (pacrob) [reply geth/utils/validation.py:22]: there's more code than just the list of kwargs that relies on autodag. I'll do a separate PR.

[the remark this answers is not in the document]
Comment (fselmo) [reply geth/utils/validation.py:22]: > there's more code than just the list of kwargs that relies on autodag. I'll do a separate PR.

I've removed it in #206

**Review by kclowes — APPROVED** [review]
Got it. Good old copilot. LGTM!

**Pull request merged**